File: src/converter/plugins/convertConditionalType.ts
import ts, {ConditionalTypeNode} from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {isPossiblyNullableType} from "./NullableUnionTypePlugin.js";

export const convertConditionalType = createSimplePlugin((node, context, render) => {
    if (!ts.isConditionalTypeNode(node)) return null

    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    const isNullableTrue = isPossiblyNullableType(node.trueType, context)
    const isNullableFalse = isPossiblyNullableType(node.falseType, context)
    const isNullable = isNullableTrue || isNullableFalse

    return `Any${isNullable ? "?" : ""} /* ${typeScriptService?.printNode(node)} */`;
})

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertConditionalType.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node
import typescript.isConditionalTypeNode

val convertConditionalType = createSimplePlugin plugin@{ node: Node, context, _ ->
    if (!isConditionalTypeNode(node)) return@plugin null

    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    val isNullableTrue = isPossiblyNullableType(node.trueType, context)
    val isNullableFalse = isPossiblyNullableType(node.falseType, context)
    val isNullable = isNullableTrue || isNullableFalse

    "Any${if (isNullable) "?" else ""} /* ${typeScriptService?.printNode(node)} */"
}
